components/resolved_ts/src/advance.rs

```diff
@@ -135,129 +137,140 @@ impl<E: KvEngine> AdvanceTsWorker<E> {
-                    cdc_clients
-                        .lock()
-                        .unwrap()
-                        .insert(store_id, TikvClient::new(channel));
-                }
-                let client = cdc_clients.lock().unwrap().get(&store_id).unwrap().clone();
-                let mut req = CheckLeaderRequest::default();
-                req.set_regions(regions.into());
-                req.set_ts(min_ts.into_inner());
-                let res = box_try!(
-                    tokio::time::timeout(
-                        Duration::from_millis(DEFAULT_CHECK_LEADER_TIMEOUT_MILLISECONDS),
-                        box_try!(client.check_leader_async(&req))
-                    )
-                    .await
-                );
-                let resp = box_try!(res);
-                Result::Ok((store_id, resp))
+                send_check_leader_request(
+                    store_id,
+                    regions,
+                    pd_client,
+                    security_mgr,
+                    env,
+                    tikv_clients,
+                    min_ts,
+                )
+                .await
             }
-        });
-        let resps = futures::future::join_all(stores).await;
-        resps
-            .into_iter()
-            .filter_map(|resp| match resp {
-                Ok(resp) => Some(resp),
-                Err(e) => {
-                    debug!("resolved-ts check leader error"; "err" =>?e);
-                    None
-                }
-            })
-            .map(|(store_id, resp)| {
-                resp.regions
-                    .into_iter()
-                    .map(move |region_id| (store_id, region_id))
-            })
-            .flatten()
-            .for_each(|(store_id, region_id)| {
+            .boxed()
+        })
+        .collect();
+    for _ in 0..store_count {
+        let (res, _, remains) = select_all(stores).await;
+        stores = remains;
+        if let Ok((store_id, resp)) = res {
+            for region_id in resp.regions {
                 resp_map.entry(region_id).or_default().push(store_id);
-            });
-        resp_map
-            .into_iter()
-            .filter_map(|(region_id, stores)| {
-                if region_has_quorum(&region_map[&region_id], &stores) {
-                    Some(region_id)
-                } else {
-                    debug!(
-                        "resolved-ts cannot get quorum for resolved ts";
-                        "region_id" => region_id,
-                        "stores" => ?stores,
-                        "region" => ?&region_map[&region_id]
-                    );
-                    None
+                if region_has_quorum(&region_map[&region_id], &resp_map[&region_id]) {
+                    valid_regions.insert(region_id);
                 }
-            })
-            .collect()
+            }
+        }
+        if valid_regions.len() == regions.len() {
```

Comment:
```suggestion
        if valid_regions.len() == regions.len() {
            // break here because all regions have quorum,
            // so there is no need waiting for other stores to respond.
```